couchcopy

```diff
@@ -59,41 +59,69 @@ async def subprocess(*args, shell=False, input=None):
                     f'{stdout}, stderr: {stderr}')
 
 
-async def backup(hostname, path, output, reuse_dir=None, tmp_dir=None,
-                 nodes_names=None):
-    tmp_dir = TemporaryDirectory(prefix='couchcopy-', dir=tmp_dir)
-    tmp_path = reuse_dir or tmp_dir.name
-    if not os.path.exists(tmp_path + '/data'):
-        os.mkdir(tmp_path + '/data')
-
-    with open(tmp_path + '/metadata.yaml', 'w') as f:
-        yaml.dump({
-            'backup': {
-                'date': datetime.now().isoformat(),
-                'source': {'nodes-names': nodes_names or []},
-            },
-            'couchcopy': {'version': __version__},
-        }, f, default_flow_style=False)
-
-    # rsync is used to copy from remote machine, but also to save files
-    # in a specific order, and to avoid using `tar` on files that are used in
-    # parallel by a running CouchDB.
-    print('[rsync...]')
-    hostname = hostname + ':' if hostname != 'localhost' else ''
-    await subprocess('rsync', '--del', '--ignore-missing-args',
-                     '-av', hostname + path + '/.shards', tmp_path + '/data')
-    await subprocess('rsync', '--del', '-av', hostname + path + '/_dbs.couch',
-                     tmp_path + '/data')
-    await subprocess('rsync', '--del', '--ignore-missing-args',
-                     '-av', hostname + path + '/shards', tmp_path + '/data')
-
-    print('[tar...]')
-    await subprocess('tar', '-I', 'pigz', '-cf', output, '-C', tmp_path,
-                     'metadata.yaml', 'data')
-    # For debugging purposes:
-    # if tmp_dir:
-    #     shutil.rmtree('/tmp/couchcopy', True)
-    #     shutil.copytree(tmp_dir.name, '/tmp/couchcopy')
+async def backup(hostname, path, output, nodes_names=None):
+    metadata = yaml.dump({
+        'backup': {
+            'date': datetime.now().isoformat(),
+            'source': {'nodes-names': nodes_names or []},
+        },
+        'couchcopy': {'version': __version__},
+    }, default_flow_style=False).encode()
+
+    print('[tar+ssh+gzip...]')
+
+    ssh_if_needed = ('ssh', hostname, '--') if hostname != 'localhost' else ()
+    tempdir = await subprocess(*ssh_if_needed,
+                               'mktemp', '-d', '-t', 'couchcopy-XXXXX')
+
+    try:
+        tempdir = tempdir.strip()
+        await subprocess(*ssh_if_needed,
+                         'dd', f'of={tempdir}/metadata.yaml', 'status=none',
+                         input=metadata)
```

Comment:
The `metadata.yaml` file to put inside the archive now needs to be created on the source machine. It would be cool to add it afterwards on the target machine, but this would require to untar it on disk (I searched and there is no way to do it "on the fly") which would defeat the performance improvement...